shipyard_blueprints/smartsheets/shipyard_smartsheets/cli/download.py

```diff
@@ -0,0 +1,156 @@
+    parser.add_argument("--sheet-id", dest="sheet_id", required=True)
+    parser.add_argument("--destination-file-name", dest="file_name", required=True)
+    parser.add_argument(
+        "--destination-folder-name", dest="folder_name", required=False, default=""
+    )
+    parser.add_argument(
+        "--file-type",
+        dest="file_type",
+        required=False,
+        default="csv",
+        choices={"xlsx", "csv"},
+    )
+    return parser.parse_args()
+
+
+def connect(logger: logging.Logger, smartsheet: smartsheet.Smartsheet):
+    try:
+        test = smartsheet.Users.get_current_user()
+    except Exception as e:
+        logger.error("Error in connecting to Smartsheet")
+        logger.error(str(e))
+        return 1
+    else:
+        return 0
+
+
+def get_logger():
+    logger = logging.getLogger("Shipyard")
+    logger.setLevel(logging.DEBUG)
+    # Add handler for stderr
+    console = logging.StreamHandler()
+    console.setLevel(logging.DEBUG)
+    # add specific format
+    formatter = logging.Formatter(
+        "%(asctime)s - %(name)s - %(levelname)s -%(lineno)d: %(message)s"
+    )
+    console.setFormatter(formatter)
+    logger.addHandler(console)
+    return logger
+
+
+def is_valid_sheet(smart: smartsheet.Smartsheet, sheet_id: str) -> bool:
+    try:
+        response = smart.Sheets.get_sheet(sheet_id, page_size=1)
+        if isinstance(response, smart.models.error.Error):
+            return False
+    except Exception as e:
+        return False
+    else:
+        return True
+
+
+def flatten_json(json_data: Dict[Any, Any]) -> Dict[str, List[Any]]:
+    try:
+        row_count = json_data["totalRowCount"]
+        columns = json_data["columns"]  # this will be a list of all the columns
+        # set up a quick lookup table for for column id and column name
+        lookup = {c["id"]: c["title"] for c in columns}
+
+        ret_dict = {}
+        # setup the keys for return dictionary and initialize and empty list as the value
+        for k in lookup.values():
+            ret_dict[k] = []
+
+        # get the row json
+        rows = json_data["rows"]
+
+        # iterate through each row and grab the cell
+        for row in rows:
+            cells = row["cells"]
+            for cell in cells:
+                col_id = cell["columnId"]
+                value = cell["value"]
+                col_name = lookup.get(col_id)
+                # update the return dictionary
+                ret_dict.get(col_name).append(value)
+    except Exception as e:
+        raise ExitCodeException(
+            message=f"Error in parsing json data: {str(e)}",
+            exit_code=ss.EXIT_CODE_DOWNLOAD_ERROR,
+        )
+
+    else:
+        return ret_dict
+
+
+def main():
+    args = get_args()
+    logger = get_logger()
+    try:
+        # establish client and check if access token is valid
+        smart = smartsheet.Smartsheet(args.access_token)
```

Comment:
looks like you're only using the sdk for validating inputs. if you tried 
```
response = requests.get(f"https://api.smartsheet.com/2.0/sheets/{sheet_id}", 
headers={"Authorization": f"Bearer {token}", "Accpet": "text/csv"})
```
would the response have specific error messaging? 
looking at https://smartsheet.redoc.ly/?_ga=2.171137568.1857487236.1698977273-1274715550.1698071792&_gl=1*vyq4gn*_ga*MTI3NDcxNTU1MC4xNjk4MDcxNzky*_ga_ZYH7XNXMZK*MTY5ODk3NzI3My41LjAuMTY5ODk3NzI3My42MC4wLjA.#section/Error-Code-List

you can remove the validation and handle the error based on status codes

i would recommend doing 
```
def get_sheet_data(sheet_id):
              response = requests.get(f"https://api.smartsheet.com/2.0/sheets/{sheet_id}", 
              headers={"Authorization": f"Bearer {token}", "Accpet": "text/csv"}
              )
              if response.ok:
                    return response.json()
               else:
                    if response.status_code ==401:
                         raise invalid credentials response.text
                     elif response.status_code == 404:
                          raise invalid sheet id response text.
                     else:
                          response.raise_for_status
```
depending on how accurate their documentation is you may be able to handle the more specific error code `Smartsheet errorCode`

Reply:
So I’m not just using the SDK for validating inputs, but actually for uploading, deleting, and fetching as well. Getting the sheet data through the api is trivial (which is why in the download blueprint, I use the api), but the value in the sdk came when appending and replacing existing sheets. [](edited-with-Axolo)

Reply:
this comment was specifically for `shipyard_blueprints/smartsheets/shipyard_smartsheets/cli/download.py`. I am only seeing `smart` referenced in `connect` and `is_valid_sheet`

Reply:
ahh I see. Slack was confusing me then. Yes that is valid, I pretty much just copied that `is_valid_sheet` from the upload.py [](edited-with-Axolo)